contracts/src/registry/PermissionedRegistry.sol

```diff
@@ -2,55 +2,264 @@
+    uint256 public constant ROLE_SET_FLAGS_ADMIN = ROLE_SET_FLAGS << 128;
+
+    uint256 public constant MAX_EXPIRY = type(uint64).max;
+    uint96 public constant FLAGS_MASK = 0xffffffff; // 32 bits
+
+    constructor(IRegistryDatastore _datastore, RegistryMetadata _metadata) BaseRegistry(_datastore) MetadataMixin(_metadata) {
+        _grantRoles(ROOT_RESOURCE, ALL_ROLES, _msgSender());
+
+        if (address(_metadata) == address(0)) {
+            _updateMetadataProvider(new SimpleRegistryMetadata());
+        }
+    }
 
-    constructor(IRegistryDatastore _datastore) BaseRegistry(_datastore) {
+    function uri(uint256 tokenId ) public view override returns (string memory) {
+        return tokenURI(tokenId);
     }
 
-    function _setFlags(uint256 tokenId, uint96 _flags)
-        internal
-        withSubregistryFlags(tokenId, FLAG_FLAGS_LOCKED, 0)
-        returns(uint96 newFlags)
+    function ownerOf(uint256 tokenId)
+        public
+        view
+        virtual
+        override(ERC1155Singleton, IERC1155Singleton)
+        returns (address)
     {
-        (address subregistry, uint96 oldFlags) = datastore.getSubregistry(tokenId);
-        newFlags = oldFlags | (_flags & FLAGS_MASK);
-        if (newFlags != oldFlags) {
-            datastore.setSubregistry(tokenId, subregistry, newFlags);
+        (, uint96 oldFlags) = datastore.getSubregistry(tokenId);
+        uint64 expires = _extractExpiry(oldFlags);
+        if (expires < block.timestamp) {
+            return address(0);
         }
+        return super.ownerOf(tokenId);
     }
 
-    function setSubregistry(uint256 tokenId, IRegistry registry)
+    function register(string calldata label, address owner, IRegistry registry, address resolver, uint96 flags, uint256 roleBitmap, uint64 expires, string memory _uri)
+        public
+        onlyRootRoles(ROLE_REGISTRAR)
+        returns (uint256 tokenId)
+    {
+        tokenId = _computeTokenId(label, flags);
+        flags = _computeFlags(flags, expires);
+
+        (, uint96 oldFlags) = datastore.getSubregistry(tokenId);
+        uint64 oldExpiry = _extractExpiry(oldFlags);
+        if (oldExpiry >= block.timestamp) {
+            revert NameAlreadyRegistered(label);
+        }
+
+        if (expires < block.timestamp) {
+            revert CannotSetPastExpiration(expires);
+        }
+
+        // if there is a previous owner, burn the token
+        address previousOwner = super.ownerOf(tokenId);
+        if (previousOwner != address(0)) {
+            _burn(previousOwner, tokenId, 1);
+        }
+
+        _mint(owner, tokenId, 1, "");
+        _grantRoles(tokenIdResource(tokenId), roleBitmap, owner);
+
+        datastore.setSubregistry(tokenId, address(registry), flags);
+        datastore.setResolver(tokenId, resolver, flags);
+
+        if (bytes(_uri).length > 0) {
+            metadataProvider.setTokenUri(tokenId, _uri);
+            emit URI(_uri, tokenId);
+        }
```

Comment:
Aside from the security issue, I don't think we need per-token URIs in general. An interface that takes the token ID and generates it based on an algorithm should work fine.

Reply:
Well we have support for any scheme. That's the point the Registry metadata implementations. Are you saying to assume we'll never externally set the uri per tokenId?

Reply:
I can't see a need to set per-token URIs in the root, .eth, or user registries. So I think for now we can leave them out. A subclass of this could always introduce them if needed.